### ISSUE TITLE
Add Automatic Version Bump Workflow for `dev` Branches

### ISSUE THREAD

**Issue by tn3rb** [eventespresso/event-espresso-core#3312]
Currently changes made to the EE Core or add-on `dev` branches do not increment the plugin version string. As we get more and more into automating our CI/CD pipeline, we should think about bumping the versions for out development branches.

I'm thinking that our `dev` branches could set their release type to `alpha`  (plz see https://github.com/eventespresso/event-espresso-core/pull/3219) and then bump the build number upon every push.

We will also want to have some sort of mechanism for bumping the version string after merges are made to `master`. So for example,

- if `master` is at version `1.0.0.rc.012` 
- and the `dev` branch is at `1.0.1.alpha.123` 
- then merging `dev` into `master` should result in:
    - `master` being bumped to `1.0.1.rc.001` 
    - `dev` being bumped to `1.0.2.alpha.001` 

Similarly, we might want to have a HOTFIX branch that can also merge changes into master and trigger the appropriate version bumps.

**Comment by manzoorwanijk** [comment]
Would that mean the two branches (`master`/`dev`) would maintain their own versions? Wouldn't that result in conflicts during merge? 🤔

**Comment by tn3rb** [comment]
Yes those two branches would be maintaining separate versions but it's not easily avoidable since there will always be the potential need to implement hotfixes. Since hotfixes do not travel down the regular development pipeline, at some point you will encounter a conflict with the versioning.

Currently, testing is performed on FET (feature) or BUG branches prior to being merged into master, with the intent that master is "ready for release" at any moment. Of course, ideally we would have everything covered with e2e testing so that manual testing wasn't necessary, but it's going to be a long time before we get ALL of our applications covered.

Ultimately, our primary goal is to have a main/master branch that is "ready for release" at any moment.

WHY?

Because if a critical bug is discovered in the wild, then we need to produce and release a hotfix as soon as possible. This means that master can NOT have code in it that has not been tested already.

So this means that we can not have untested code being deployed directly into master branch. It needs to go into some other branch where it can be tested prior to merging into master (OR... release branches are cut from master and THEN tested, but that's essentially the exact same thing just with different branch names). That makes me realize that we will also need another intermediary branch between `dev` and `master` to merge into so that automatic deployments can continue to the `dev` branch without invalidating any testing that has already occurred.

With regards to conflicts due to version string changes; since that will only occur in the plugin main file, those conflicts can be dealt with in a shell script via something like:

```
git checkout --theirs espresso.php
```

So to reiterate we're looking for something like:

_`--manual-merge-->`_ **` barista master `**

_`--auto-deploy-->`_ **` dev branch `**

_`--version-bump-->`_ **` dev branch `**

`--manually-triggered-merge-->` **` release-candidate `**

_`--version-bump-->`_ **` release-candidate `**

_`--auto-triggered-->`_ **` automated e2e tests + SB `**

if ` tests fail` _`-->`_ ` priority repair (ie: we fix stat) `

**` manual Q/A testing `**

if `tests fail` _`-->`_ ` priority repair `

`--manually-triggered-merge-->` **` staging `**

_`--version-bump-->`_ **` staging `**

_`--auto-deploy-->`_ **` MENW server `**

**` manual Q/A testing `**

if `tests fail` _`-->`_ ` priority repair `

`--manually-triggered-merge-->` **` master `**

_`--version-bump-->`_ **` master `**

`--manually-triggered-release-ES-->` **` master `**

_`--auto-deploy-->`_ **` ES server `**

`--manually-triggered-release-EE-->` **` master `**

_`--auto-package-->`_ **` EE plugin prepped `**

_`--auto-deploy-->`_ **` EE plugin server `**

_`--auto-package-->`_ **` EE decaf prepped `**

_`--auto-deploy-->`_ **` WP.org plugins server `**


Hope that makes sense, it's a little difficult trying to map that all out just using text
